### PR TITLE
#17170 - DATEDIFF Week Implementation

### LINKED ISSUE #17170: please add feature `DateDiffWeek` extension for mssql
Please add feature `DateDiffWeek` DbFunctions extension for mssql

<https://github.com/aspnet/EntityFrameworkCore/blob/release/3.0/src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs>
<https://docs.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql?view=sql-server-2017>

### PR DESCRIPTION
**Pull request by ralmsdeveloper**
Resolve #17170

### CODE REVIEW

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 1]
```diff
@@ -880,6 +881,97 @@ private static bool ContainsCore(string propertyName, string searchCondition, in
...
                 : null;
 
+        /// <summary>
+        ///     Counts the number of year boundaries crossed between the startDate and endDate.
+        ///     Corresponds to SQL Server's DATEDIFF(WEEK,startDate,endDate).
+        /// </summary>
+        /// <param name="_">The DbFunctions instance.</param>
+        /// <param name="startDate">Starting date for the calculation.</param>
+        /// <param name="endDate">Ending date for the calculation.</param>
+        /// <returns>Number of year boundaries crossed between the dates.</returns>
```
Comment (btecu) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs]: Number of week not year

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 2]
```diff
@@ -880,6 +881,97 @@ private static bool ContainsCore(string propertyName, string searchCondition, in
...
+
+                if (remainingDays > 0)
+                {
+                    var calendar = CultureInfo.InvariantCulture.Calendar;
+
+                    var first = calendar
+                        .GetWeekOfYear(
+                            startDate,
+                            CalendarWeekRule.FirstFullWeek,
+                            DayOfWeek.Monday);
```
Comment (btecu) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs]: I'm not sure if it matters but in most places the first day of the week is Sunday

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 3]
```diff
@@ -880,6 +881,97 @@ private static bool ContainsCore(string propertyName, string searchCondition, in
...
+        }
+
+        /// <summary>
+        ///     Counts the number of year boundaries crossed between the startDate and endDate.
+        ///     Corresponds to SQL Server's DATEDIFF(WEEK,startDate,endDate).
+        /// </summary>
+        /// <param name="_">The DbFunctions instance.</param>
+        /// <param name="startDate">Starting date for the calculation.</param>
+        /// <param name="endDate">Ending date for the calculation.</param>
+        /// <returns>Number of year boundaries crossed between the dates.</returns>
```
Comment (btecu) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs]: Number of weeks instead of year

**Comment by ErikEJ** [comment]
DATEDIFF always uses Sunday as the first day of the week to ensure the function operates in a deterministic way.

**Comment by ralmsdeveloper** [comment]
kkk, I climbed my last test, was testing some possibilities.

I was sure it was sunday.

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs (hunk 1 above)
Comment (ralmsdeveloper) [reply src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs]: Thanks, I really missed it.

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 4]
```diff
@@ -880,6 +881,97 @@ private static bool ContainsCore(string propertyName, string searchCondition, in
                 ? (int?)DateDiffNanosecond(_, startTimeSpan.Value, endTimeSpan.Value)
                 : null;
 
+        /// <summary>
+        ///     Counts the number of year boundaries crossed between the startDate and endDate.
```
Comment (btecu) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs]: `year` -> `week`

**Comment by WeihanLi** [comment]
@ralmsdeveloper thanks for your effort, could there's a parameter to specific the the first day of week, the Monday is the first day of the week for China in most cases

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs [hunk 5]
```diff
@@ -880,6 +881,97 @@ private static bool ContainsCore(string propertyName, string searchCondition, in
...
+        /// <param name="endDate">Ending date for the calculation.</param>
+        /// <returns>Number of week boundaries crossed between the dates.</returns>
+        public static int DateDiffWeek(
+            [CanBeNull] this DbFunctions _,
+            DateTime startDate,
+            DateTime endDate)
+        {
+            checked
+            {
+                var days = (endDate.Date - startDate.Date).Days;
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs]: @ajcvickers @divega @bricelam - Are we adding client side implementation for DbFunctions which are specifically mapped to provider's function on server side? This had come up in one of the meetings.

File: test/EFCore.SqlServer.FunctionalTests/Query/DbFunctionsSqlServerTest.cs [hunk 6]
```diff
@@ -560,6 +560,23 @@ public virtual void DateDiff_Nanosecond()
             }
         }
 
+        [ConditionalFact]
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/DbFunctionsSqlServerTest.cs]: Tests:
- Use DateTime & DateTimeOffset both column types.
- Non-zero result.

**Comment by ajcvickers** [comment]
@smitpatel Note that this isn't a priority; we're not taking this for 3.0.

**Comment by smitpatel** [comment]
> could there's a parameter to specific the the first day of week, the Monday is the first day of the week for China in most cases

No. That is not possible. According to SqlServer documentation
`SET DATEFIRST` is used to configure first day of the week. DATEDIFF function has nothing to do with it. And as @ErikEJ mentioned above, it does not have any effect on DATEDIFF.
If you add such parameter then there is no SQL translation for it.

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs (hunk 5 above)
Comment (bricelam) [reply src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs]: I think we should. IMHO, working with AsEnumerable is still important

**Comment by WeihanLi** [comment]
> > could there's a parameter to specific the the first day of week, the Monday is the first day of the week for China in most cases
> 
> No. That is not possible. According to SqlServer documentation
> `SET DATEFIRST` is used to configure first day of the week. DATEDIFF function has nothing to do with it. And as @ErikEJ mentioned above, it does not have any effect on DATEDIFF.
> If you add such parameter then there is no SQL translation for it.

@shimingsg thanks for your info, got it from the docs [here](https://docs.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql?view=sql-server-2017#remarks)

> Specifying SET DATEFIRST has no effect on DATEDIFF. DATEDIFF always uses Sunday as the first day of the week to ensure the function operates in a deterministic way.

File: src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs (hunk 5 above)
Comment (smitpatel) [reply src/EFCore.SqlServer/Extensions/SqlServerDbFunctionsExtensions.cs]: @aspnet/efteam - Someone needs to verify if this client side implementation is correct for T-SQL. I tried to go down rabbit hole and fell into pit of failure.

**Comment by ralmsdeveloper** [comment]
> @ralmsdeveloper agradecimentos para seu esforço, poderia lá é um parâmetro a específico o o primeiro dia da semana, a segunda-feira é o primeiro dia da semana para China em a maioria de casos

We can't do this, sql server uses deterministically Sunday, that means the translation will not match the SQL Server result.
Besides the DATEDIFF function does not have the parameter to define what would be the first day of the week.

[https://docs.microsoft.com/pt-br/sql/t-sql/functions/datediff-transact-sql?view=sql-server-2017](https://docs.microsoft.com/pt-br/sql/t-sql/functions/datediff-transact-sql?view=sql-server-2017)

@WeihanLi 
But if you really have this need you can build a custom function for yourself.

**Comment by ralmsdeveloper** [comment]
> Need more testing as requested.

Done!

**Comment by smitpatel** [comment]
Merging to release/3.1 in #17726

@ralmsdeveloper - Thank you for contribution.

**Pull request closed without merging**